[22기 최재상] Refactoring 및 실제 서버로부터 데이터 수신하도록 수정

Higher77 opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 리팩토링 및 실제 서버로부터 데이터 수신
- 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
-StarRating.js의 중복되는 코드를 리팩토링 하였습니다.
-백엔드 서버로부터 데이터를 수신하여 실제로 데이터가 실시간 반영되도록 수정

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

고쳐야할 점이 없다면 Merge부탁드립니다!

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
[Add] Star Css기본설정

Commit:
[Add] 별점 클릭기능 구현
-원하는 별점  클릭시 해당하는 별점이 매겨집니다.
-클릭했던 별점을 다시 클릭하면 초기화가 됩니다.
-마우스 호버효과는 미구현입니다.

Commit:
[Add] 별점기능 호버,클릭,취소 기능 구현

Commit:
[Add] 컴포넌트 재활용을 위한 구조 파일구조 변경
-StarRating과 StarGraph를 형제컴포넌트로 분리하였습니다.
-재활용을 고려하여 Test.js부모 컴포넌트에서 불러와 정상작동하는지 확인했습니다.

Commit:
[Add] 가짜서버를 통한  데이터 실시간 업데이트 구현중

Commit:
[Add] 실시간 반응형 별점 그래프 구현

Commit:
[Add] 별점 그래프 CSS수정 및 최고 비율 별점 막대 색 변환

Commit:
[Add] 새로고침 후에도 평가 별점이 유지되는 것을 구현

Tonyk0901 commented on src/Components/Star/StarRating/StarRating.js:
```diff
@@ -0,0 +1,197 @@
+      .then(data => {
+        const currentRateValue = Array(5)
+          .fill(false)
+          .fill(true, 0, data.rating);
```
- 👍 👍

Tonyk0901 commented on src/Components/Star/StarRating/StarRating.js:
```diff
@@ -0,0 +1,197 @@
+        hoverRateValue: [false, false, false, false, false],
+        rateValue: prevRateValue,
+      });
+
```
- 너무 많은 부분이 동일하게 반복되고 있습니다! 달라지는 부분에 대해서만 조건을 걸어주시면 코드의 양을 더 많이 줄일 수 있을 것 같습니다!
```suggestion
    if (isClickedStarActive && isNextStarActive) {
      prevRateValue.fill(false, clickedIndex + 1);
    } else if (isClickedStarActive) {
      prevRateValue.fill(false, 0, clickedIndex + 1);    
    } else if (!isClickedStarActive) {
      prevRateValue.fill(true, 0, clickedIndex + 1);    
    }
    const rating = prevRateValue.filter(value => value).length;
    fetch('http://localhost:1313/rating/7', {
      method: 'PATCH',
      body: JSON.stringify({
        rating: rating,
      }),
      headers: { 'Content-type': 'application/json; charset=UTF-8' },
    }).then(() => {
      this.props.callApi();
    });

    this.setState({
      isHover: false,
      hoverRateValue: [false, false, false, false, false],
      rateValue: prevRateValue,
    });
```

Tonyk0901 commented on src/Components/Star/StarRating/StarRating.js:
```diff
@@ -0,0 +1,197 @@
+      });
+
+      return;
+    }
```
- 여기도 달라지는 부분에 대해서만 조건을 사용해주시면 좋을 것 같습니다!

Tonyk0901 commented on src/Components/Star/StarRating/StarRating.js:
```diff
@@ -0,0 +1,197 @@
+  };
+
+  checkIsActive = star => {
+    if (this.state.isHover) {
+      if (this.state.hoverRateValue[star]) {
+        return 'activeStar';
+      }
+
+      return 'inactiveStar';
+    }
+
+    if (this.state.rateValue[star]) {
+      return 'activeStar';
+    }
+
+    return 'inactiveStar';
```
- 결국엔 경우의 수가 2개이기 때문에, 이렇게 너무 많은 조건을 사용할 필요가 없어 보입니다! 패턴 하나 소개해드립니다! 참고해주세요!
```suggestion
    const { isHover, hoverRateValue, rateValue } = this.state;
    const isStarActive = (isHover && hoverRateValue[star]) || (rateValue[star]);
    return isStarActive ? 'activeStar' : 'inactiveStar';
```
- 조건이 복잡할 경우엔 조건을 변수에 담고 해당 변수의 이름을 조건의 내용을 알 수 있는 것으로 네이밍 해주는 것을 통해, 조건을 직접 해석하지 않아도 조건이 어떤 의미를 가지는지 파악할 수 있도록 해주시는 것이 좋습니다.

Higher77 replied:
성훈님 제가 작성한 코드에서는 isHover와 hoverRateValue[star]가 둘 다 false면 그대로 return 하면서 끝나는데
참고로 올려주신 코드에서 isHover와 hoverRateValue[star] 둘 다 false여도 rateValue[star]가 true인 경우가 나와서요.

제가 && ||을 이용해서 고쳐보려고 했는데 도저히 떠오르지가 않아서

    const { isHover, hoverRateValue, rateValue } = this.state;
    const isStarActive = rateValue[star];

    if (isHover) {
      return hoverRateValue[star] ? 'activeStar' : 'inactiveStar';
    }

    return isStarActive ? 'activeStar' : 'inactiveStar';
  

이렇게까지만 작성했습니다... ㅠㅠ;  더 좋은 방법이 있을까요?

Tonyk0901 replied:
아뇨-! 얼리 리턴을 사용해서 작성해주신 현재 코드 깔끔한 것 같습니다! LGTM ✨

Tonyk0901 reviewed (commented):
국-! 코멘트 확인해주세요 재상님-!

Commit:
[Refactor] 리팩토링중 백업

Commit:
[Refactor] 중복되는 코드 리팩토링
-StarRating.js파일  handleStarClick, handleStarMousehover, checkIsActive의 중복코드 통합 및 리팩토링

Commit:
[Modify] 백엔드로부터 직접 데이터를 받도록 fetch수정

Commit:
[Merge]  'master'브런치 feature/Star merge

Commit:
[Add] token 유무 따른 별점,그래프 변화
-token이 없을시 별점을 매길 수 없고 경고창이 나옵니다.
-token이 있을시 별점을 매길 수 있고 그래프에 반영됩니다.

Commit:
[Modify] 테스트 파일 삭제 , 별점그래프 표현방식 수정
-test.js파일을 삭제하고 Routes.js에 적힌 경로역시 삭제하였습니다.
-StarGraph에서 백분율이 0%인 경우 아예 표시하지 않는게 아니라 1%를 주도록 바꿨습니다.
